refactor(localize): use the new workspaces API for ng-add schematic

SchnWalter opened the pull request:
## PR Checklist
Please check if your PR fulfills the following requirements:

- [x] The commit message follows our guidelines: https://github.com/angular/angular/blob/master/CONTRIBUTING.md#commit
- [ ] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)


## PR Type
What kind of change does this PR introduce?

<!-- Please check the one that applies to this PR using "x". -->

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] CI related changes
- [ ] Documentation content changes
- [ ] angular.io application / infrastructure changes
- [ ] Other... Please describe:


## What is the current behavior?

Updates the `ng add @angular/localize` to use the new Workspaces API.

Issue Number: N/A

Related to PR https://github.com/angular/angular-cli/pull/17597 and https://github.com/angular/components/pull/19229


## What is the new behavior?

As the title describes, this PR updates the schematic to use the new workspaces API.


## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No


<!-- If this PR contains a breaking change, please describe the impact and migration path for existing applications below. -->


## Other information

This PR is a followup for https://github.com/angular/angular-cli/pull/17597 where I've discovered that there are some unused methods in `@schematics/angular` and removed them.

cc: @alan-agius4

SchnWalter commented on packages/localize/schematics/ng-add/index.ts:
```diff
@@ -8,53 +8,67 @@
-import {BrowserBuilderTarget, Builders, ServeBuilderTarget} from '@schematics/angular/utility/workspace-models';
+import {virtualFs, workspaces} from '@angular-devkit/core';
+import {chain, Rule, SchematicsException, Tree} from '@angular-devkit/schematics';
+import {getWorkspace} from '@schematics/angular/utility/workspace';
```
I tried replacing this `getWorkspace()` with something from `@angular-devkit/core`, but the only API I found was `workspaces.readWorkspace()`, which also requires a file path and doesn't work on the available `host: Tree`, see:

https://github.com/angular/angular-cli/tree/v10.0.0-next.3/packages/angular_devkit/core#usage-example

SchnWalter commented on packages/localize/schematics/ng-add/index_spec.ts:
```diff
@@ -155,6 +156,7 @@ export { renderModule, renderModuleFactory } from '@angular/platform-server';`;
 
   it('should not break when there are no polyfills', async () => {
     host.overwrite('angular.json', JSON.stringify({
+      version: 1,
```
Looks like the new Workspaces API requires this version flag, otherwise it throws:
```
throw new Error('Unsupported workspace format.');
```

But there is no mention of this in the https://angular.io/guide/deprecations and the codebase seems to not "require" this flag:
```ts
// NOTE: future additions could also perform content analysis to determine format/version
const workspaceFiles: Record<string, WorkspaceFormat> = {
  'angular.json': WorkspaceFormat.JSON,
  '.angular.json': WorkspaceFormat.JSON,
};
```

If this version flag will be added in the future, the new API should first display a warning and not throw an Error, it should go through a deprecation phase.

petebacondarwin replied:
@alan-agius4 ^^^

alan-agius4 replied:
angular.json must always have a version so this is LGTM

SchnWalter replied:
Thank you.

I've now updated the `Does this PR introduce a breaking change?` for this MR.

petebacondarwin commented on packages/localize/schematics/ng-add/index.ts:
```diff
@@ -8,53 +8,70 @@
   return (host: Tree) => {
-    // Get all known polyfills for browser builders on this project.
-    const optionValues = getAllOptionValues<string>(host, projectName, builderName, optionName);
+    const optionValues = getRelevantFilePathsForProject(project, builderName, optionName);
```
NIT: Given all the other renames in this file, I would expect this to be something other than `optionValues`. Perhaps `filePaths`?

petebacondarwin replied:
Or `fileList`?

SchnWalter replied:
I forced pushed a new version with 2 changes:
* renamed `getRelevantFilePathsForProject` to `getFileListForRelevantTargetDefinitions`
* renamed `optionValues` to `fileList`

petebacondarwin reviewed (approved):
LGTM - a few nits.

alan-agius4 commented on packages/localize/schematics/ng-add/index.ts:
```diff
@@ -8,55 +8,72 @@
+function getRelevantTargetDefinitions(
+    project: workspaces.ProjectDefinition, builderName: Builders): workspaces.TargetDefinition[] {
+  const definitions: workspaces.TargetDefinition[] = [];
+  for (const target of project.targets.values()) {
```
Not: you shouldn’t be needed `.values` As target is an iterable.

SchnWalter replied:
Changed to `for (const [, target] of project.targets) {...}` and rebased onto master.

petebacondarwin replied:
I'm not sure that this is better :-) but happy to go with it.

SchnWalter replied:
I'm not sure exactly how you would like to change this, could you please give an example.

If the `values()` method is not used, the iterator returns an `IterableIterator<[string,TargetDefinition]>` and to access the `TargetDefinition`, you either need to access it via an array index or by using a destructuring assignment, and I looked over the Angular CLI codebase and saw that there are 10 destructuring assignments that ignore some of the returned values and there were none where the destructuring assignment kept all values but didn't use some of them.

petebacondarwin replied:
I'm not proposing a change. Just pointing out that although `targets` is iterable, it is an iterable of "entries", which results in the ugly `[, target]` destructuring.
Since `targets` is a `Map` (internally) then `targets.value()` is not costly to call, since it doesn't actually create a new array.
If I was pushed to make a suggestion, then I would do something like:

```ts
project.targets.forEach(target => {
  if (target.builder === builderName) {
    definitions.push(target);
  }
});
```

SchnWalter replied:
Interesting, I didn't know that `forEach()` will provide `TargetDefinition, string` and not `[string,TargetDefinition], number`.

Thanks. I've changed the code to match your suggestion.

SchnWalter committed:
refactor(localize): use the new workspaces API for ng-add schematic

Updates the @angular/localize ng-add schematic to use the new workspaces API and removes dependency on private APIs.
